docs/proposals/odo-deploy.md

```diff
@@ -0,0 +1,152 @@
+- The implementation would be under the experimental flag.
+- Only supported for Devfile v2.0.0 components.
+- Only supported for Kubernetes/ OpenShift targets.
+- `odo deploy` will not make any changes to the user's project.
+
+#### odo deploy 
+This command will build a container image for the application and deploy it on the target Kubernetes environment. 
+
+Flags:
+ - `--tag`: The tag to be used for the built application container image - `<registry>/<org>/<name>:<tag>` (mandatory).
+ - `service-account`: The service account for running privileged containers and push access to the image registry.
+ - `--credentials`: The credentials needed to push the image to the container image registry (mandatory).
+
+#### odo deploy delete
+This command will delete any resources created by odo deploy.
+
+### Detailed design: 
+
+### Devfile
+
+For the initial implementation, we could use devfile v2.0.0 and capture basic outer-loop information as `attributes` on the devfile. `odo` could look for  specific keys, while other tools like Che could ignore them.
+
+For example: 
+```
+attributes:
+    build-dockerfile: <URI>
+    deployment-manifest: <URI>
+```
+
+This should not need any change to the devfile v2 schema or the parser.
+
+### Dockerfile
+This could be any valid dockerfile.
+
+### Deployment Manifest
+The deployment manifest could be templated to help with replacing key bits of information:
+- PROJECT_NAME
+- CONTAINER_IMAGE
+- PORT
+
+For example: 
+```
+apiVersion: app.stacks/v1beta1
+kind: RuntimeComponent
+metadata:
+    name: PROJECT_NAME
+spec:
+    applicationImage: CONTAINER_IMAGE 
+    service:
+        type: ClusterIP
+        port: PORT
+    expose: true
+    storage:
+        size: 2Gi
+        mountPath: "/logs"
+```
+
+### odo deploy
+This command will perform the following actions:
+
+#### Input validation
+- Check if the devfile  is v2.0.0 and that it specifies the expected outer-loop attributes. 
+    - If not provided, display a meaningful error message.
+- Validate all arguments passed by the user. 
+    - If argument values are invalid, display a meaningful error message.
+
+#### Build
+- Start a new pod with a build container (privileged container with `buildah:stable` image).
+- Mount the image registry credentials passed by the user.
+- Copy the application source code into the build container.
+- Fetch the dockerfile using URI in the attributes of the devfile and make it available in the container.
+- Execute `buildah bud` command using src code, dockerfile and tag information.
+- Push the built image using the tag and credentials.
+- Delete the build container and the pod.
+
+#### Deploy
+- Delete any existing deployments (if the user runs `odo deploy` multiple times).
+- Fetch the deployment manifest using URI in the attributes of the devfile. and make it available in the container.
+- Replace templated text in the deployment manifest with relevant values:
+    - PROJECT_NAME: name of odo project
```

Comment:
I think this brings me back to the point I made earlier about `metadata.name`. IMO, `PROJECT_NAME` can't be in the `metadata.name`. Maybe, `metadata.generateName`. But that depends on whether my understanding in the previous comment is correct.

Reply:
Sure, I'm not an expert on this either. But, the point is that the devfile creator will be able to provide a manifest with these placeholders and application developers don't need to worry about it.

Reply:
If we foresee multiple projects from same user with same name which need to be deployed, we might have to add a  random suffix to make them unique. Is that what you are thinking of?

Reply:
@neeraj-laad I think it's safe to leave this part for implementation detail. I'm giving an example below only to explain the scenario that I'm concerned about. I'm sure that the dev who's implementing this will have this scenario covered or would encounter it sooner rather than later. :wink: 

The "random suffix" part is what I'm thinking of for a scenario like below:
1. I have an OpenShift/k8s project called `myproject`.
1. I am an enterprise developer working on a microservice application that's got:
    - nodejs frontend
    - java backend
    - postgresql database
1. To have a working end product, I need to deploy all the above three in same OpenShift/k8s namespace.
1. In the context of odo, there's a unique component for nodejs and java part of above architecture while postgresql is a service.
1. If the manifest will have `metadata.name` value set as `PROJECT_NAME` for both nodejs and java components, there will be a conflict in creating a k8s resource and we'll be able to create only one resource with the value of `PROJECT_NAME` value. So, it might help to have a random suffix (`metadata.generateName` can be handy.)

Reply:
Thanks for explaining this @dharmit 

With PROJCT_NAME, I was referring to the name of this specific component / micro service (node-frontend, or java-backend etc. I think the terminology might be causing the confusion.

Is there a specific name in odo for each micro-service/component the user is developing? Is COMPONENT_NAME a better term?

Reply:
Ah, now it makes sense. Yes, "COMPONENT_NAME" is the term on odo side of the things.

Reply:
I'll update the PR to reflect that.. Apologies for the confusion.